### PR TITLE
Sema: Allow non-final classes to satisfy properties and subscripts with covariant Self

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Hopefully, whatever complications were impeding this in 926e3711d0dc are long gone.

Bonus: now we can use protocols with such requirements as types (iff the requirement was the reason we previously couldn't).

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
Should I be adding some execution tests?

**Review by slavapestov — COMMENTED** [review]
Sounds good! Runtime tests would be nice to make sure the generated witness thunk is valid and recovers the correct Self type.

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

File: test/SILGen/dynamic_self.swift [hunk 1]
```diff
@@ -83,17 +93,65 @@ func testExistentialDispatch(p: P) {
...
+// CHECK:   [[PCOPY_ADDR_1:%[0-9]+]] = alloc_stack $@opened([[N]]) P
+// CHECK:   copy_addr [[PCOPY_ADDR]] to [initialization] [[PCOPY_ADDR_1]] : $*@opened([[N]]) P
+// CHECK:   [[P_P_GETTER:%[0-9]+]] = witness_method $@opened([[N]]) P, #P.p!getter : {{.*}}, [[PCOPY_ADDR]]{{.*}} : $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> @out τ_0_0
+// CHECK:   [[P_RESULT_ADDR2:%[0-9]+]] = init_existential_addr [[P_RESULT]] : $*P, $@opened([[N]]) P
+// CHECK:   apply [[P_P_GETTER]]<@opened([[N]]) P>([[P_RESULT_ADDR2]], [[PCOPY_ADDR_1]]) : $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> @out τ_0_0
+// CHECK:   destroy_addr [[PCOPY_ADDR_1]] : $*@opened([[N]]) P
+// CHECK:   destroy_addr [[P_RESULT]] : $*P
+// CHECK:   dealloc_stack [[PCOPY_ADDR_1]] : $*@opened([[N]]) P
+// CHECK:   dealloc_stack [[P_RESULT]] : $*P
+  _ = p.p
```
Comment (AnthonyLatsis) [inline test/SILGen/dynamic_self.swift:108]: If this output happens to be sane, why do we need `PCOPY_ADDR_1` in addition to `PCOPY_ADDR`?

File: test/SILGen/dynamic_self.swift [hunk 2]
```diff
@@ -83,17 +93,65 @@ func testExistentialDispatch(p: P) {
...
+// CHECK: [[CP_ADDR:%[0-9]+]] = open_existential_ref [[CP]] : $CP to $@opened([[N:".*"]]) CP
+// CHECK: [[CP_ADDR_1:%[0-9]+]] = copy_value [[CP_ADDR]] : $@opened([[N]]) CP
+// CHECK: [[CP_BORROWED:%[0-9]+]] = begin_borrow [[CP_ADDR_1]] : $@opened([[N]]) CP
+// CHECK: [[CP_P_GETTER:%[0-9]+]] = witness_method $@opened([[N]]) CP, #CP.p!getter : {{.*}}, [[CP_ADDR]]{{.*}} : $@convention(witness_method: CP) <τ_0_0 where τ_0_0 : CP> (@guaranteed τ_0_0) -> @owned τ_0_0
+// CHECK: [[APPLY_RESULT:%[0-9]+]] = apply [[CP_P_GETTER]]<@opened([[N]]) CP>([[CP_BORROWED]]) : $@convention(witness_method: CP) <τ_0_0 where τ_0_0 : CP> (@guaranteed τ_0_0) -> @owned τ_0_0
+// CHECK: end_borrow [[CP_BORROWED]] : $@opened([[N]]) CP
+// CHECK: destroy_value [[CP_ADDR_1]] : $@opened([[N]]) CP
+// CHECK: [[RESULT_EXISTENTIAL:%[0-9]+]] = init_existential_ref [[APPLY_RESULT]] : $@opened([[N]]) CP : $@opened([[N]]) CP, $CP
+// CHECK: destroy_value [[RESULT_EXISTENTIAL]] : $CP
+  _ = cp.p
```
Comment (AnthonyLatsis) [inline test/SILGen/dynamic_self.swift:143]: Same here, but with a borrow.

**Review by AnthonyLatsis — COMMENTED** [review]
I've extended some silgen tests to cover subscripts and properties as well. My understanding of the output is rather sketchy, but there *is* one thing that looks suspicious to my eye.

File: test/SILGen/dynamic_self.swift (hunk 1 above)
Comment (slavapestov) [reply test/SILGen/dynamic_self.swift:108]: I think it's just SILGen inserting an unnecessary copy. It does that sometimes.

**Review by slavapestov — APPROVED** [review]

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test and merge

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test Windows

**Pull request merged**